cppser/dbman.h

```diff
@@ -0,0 +1,245 @@
+        static void set_hms(std::string& str, int ticks);
+
+        Clock();
+
+        virtual ~Clock() noexcept = default;
+
+        // This is the interface subclasses need to implement
+        virtual std::string operator () () = 0;
+
+        // Given a string of 'xx:xx:xx', returns the second it is in a day
+        // For example, '07:00:00' -> 3600 * 7
+        static int str2time(const std::string& timestr);
+
+        // Reverse of above.
+        // Format: 07:20:00
+        static std::string time2str(int ticks);
+    };
+
+    // This clock returns the current time in the format
+    // 2021-12-31 21:50:34.1234567
+    class CurrentClock : public Clock {
+    public:
+        [[nodiscard]] virtual std::string operator() () override;
+
+        virtual ~CurrentClock() = default;
+    };
+
+    class IncrementalClock : public Clock {
+    private:
+        // ticks since start of day
+        int mTicks;
+    public:
+        IncrementalClock();
+
+        virtual std::string operator() () override;
+    };
+
+    // This clock returns a random time in a specified region
+    class RandomClock : public Clock {
+    private:
+        std::function<int()> mRandTime;
+    public:
+        // The int argument can be obtained by calling str2time()
+        RandomClock(int start, int end);
+
+        virtual ~RandomClock() noexcept = default;
+
+        virtual std::string operator() () override;
+    };
+
+    // This clock returns the timestr for a particular time, with the T character in place.
+    // Intended for API manipulation, not for sign in purposes!
+    class GivenTimeClock {
+    public:
+        // The time is given as the tick.
+        [[nodiscard]] std::string operator() (int ticks);
+    };
+
+    // This structure represents a Lesson with all its data included
+    struct LessonInfo {
+        // Endtime of the session.
+        int endtime;
+        // Lesson ID as a string obtained from DB
+        std::string id;
+        // Anpai id, required for accessing the get_stu_new API.
+        int anpai;
+    };
+
+    // Returns a vector of today's lessons' information
+    // Expects that the database now contains the required info
+    std::vector<LessonInfo> get_lesson(Connection& conn);
+
+    // Returns the machine's ID
+    std::string get_machine(Connection& conn);
+
+    // This represents a student, with his or her name and id.
+    struct Student {
+        // name: UTF-8 encoded string.
+        // id: student's id
+        std::string name, id;
+    };
+
+    // Gets a vector of Students who are still absent.
+    std::vector<Student> report_absent(Connection& conn, const std::string& lesson_id);
+
+    // Writes records to the database using the given clock for the given card numbers
+    // for the given lesson. (Whew)
+    // If access to the database is interrupted, will retry.
+    // This is a little slow, so you can use async() to launch it.
+    // Returns the SQLite error code.

```

Comment:
This return value is strange. After `SQLError` has been implemented, we can return `void`.
The transaction makes this function sort of an atomic operation.